src/System.Threading.Tasks.Extensions/src/System/Runtime/CompilerServices/AsyncBuilderAttribute.cs

```diff
@@ -5,14 +5,20 @@
 namespace System.Runtime.CompilerServices
 {
     /// <summary>
-    /// Indicates the async method builder that should be used by the C# compiler to build
-    /// the attributed type when used as the return type of an async method.
+    /// Indicates the async method builder that should be used by a language compiler to
+    /// build the attributed type when used as the return type of an async method.
     /// </summary>
     [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Interface, Inherited = false, AllowMultiple = false)]
-    internal sealed class AsyncBuilderAttribute : Attribute
+    public sealed class AsyncBuilderAttribute : Attribute
```

Comment:
I didn't test enums+delegates, but I'm pretty sure they can be made tasklike.

Detection is entirely name-based. Or more precisely, it looks for an attribute constructor named `System.Runtime.CompilerServices.AsyncBuilderAttribute[.ctor]` and which takes a single constructor whose type is `System.Type`. (not sure if it specifically looks for `System.Type` coming from a particular assembly but I bet it does).

For extending an existing interface -- we alas don't have any good way to do that. Our two proposals for how to identify a tasklike type were (1) an attribute (which requires you to modify the original definition), or (2) a static method on the tasklike called `CreateAsyncMethodBuilder` (which requires you to modify the original definition and doesn't work with interfaces). We also imagined (3) an extension method which gets passed a dummy null "this" argument, but didn't like it.

Our thinking was that for most purposes, it makes sense for the original author of the type to mark it tasklike. The only folks who will suffer from this decision are you Bart, and me -- the two people in the world who want to experiment with making existing things tasklike, for purpose of experimentation, but without control of the original type.

We were keen on allowing [AsyncBuilder(...)] attribute to be placed on async *methods*. This would allow to experiment. I'm hoping this feature could get into C# 7.1.

For `IAsyncOperation<T>` I reached out to WinRT folk but didn't get much interest. Partly that's because my recommendation was *not* to support it right now. That's because we can't support `IAsyncOperation.Cancel()` method -- that would require some way for the async method instance to communicate with the builder. Hence my wish for the `async` contextual keyword. I think that once this comes about, then we should make the change.